cmd/srcd/cmd/root.go

```diff
@@ -88,37 +88,86 @@ func initConfig() {
 
 var logMsgRegex = regexp.MustCompile(`.*msg="(.+)"`)
 
-func logAfterTimeout(header string) chan struct{} {
-	logs, err := daemon.GetLogs()
-	if err != nil {
-		logrus.Errorf("could not get logs from server container: %v", err)
+func logAfterTimeout(msg string, timeout time.Duration) func() {
+	d := &Defered{
+		Timeout: timeout,
+		Msg:     msg,
 	}
 
-	started := make(chan struct{})
-	go func() {
-		defer logs.Close()
+	return d.Print()
+}
+
+func logAfterTimeoutWithSpinner(msg string, timeout time.Duration, spinnerInterval time.Duration) func() {
+	d := &Defered{
+		Timeout:         timeout,
+		Msg:             msg,
+		Spinner:         true,
+		SpinnerInterval: spinnerInterval,
+	}
+
+	return d.Print()
+}
 
-		select {
-		case <-time.After(3 * time.Second):
-			logrus.Info(header)
+func logAfterTimeoutWithServerLogs(msg string, timeout time.Duration) func() {
+	inputFn := func(stop <-chan bool) <-chan string {
+		logs, err := daemon.GetLogs()
+		if err != nil {
+			logrus.Errorf("could not get logs from server container: %v", err)
+			return nil
+		}
+
+		ch := make(chan string)
+		go func() {
+			defer logs.Close()
 			scanner := bufio.NewScanner(logs)
-			for scanner.Scan() {
-				select {
-				case <-started:
-					return
-				default:
-					match := logMsgRegex.FindStringSubmatch(scanner.Text())
-					if len(match) == 2 {
-						logrus.Info(match[1])
-					}
+			for msg := range scanWithTimeout(scanner, 250*time.Millisecond, stop) {
+				match := logMsgRegex.FindStringSubmatch(msg)
+				if len(match) == 2 {
+					ch <- match[1]
 				}
 			}
+
+			close(ch)
 			if err := scanner.Err(); err != nil && err != context.Canceled {
 				logrus.Errorf("can't read logs from server: %s", err)
 			}
-		case <-started:
+		}()
+
+		return ch
+	}
+
+	d := &Defered{
+		Timeout: timeout,
+		Msg:     msg,
+		InputFn: inputFn,
+	}
+
+	return d.Print()
+}
+
+func scanWithTimeout(s *bufio.Scanner, timeout time.Duration, stop <-chan bool) <-chan string {
+	texts := make(chan string)
+	c := make(chan bool)
+
+	scan := func() {
+		c <- s.Scan()
+	}
+
+	go func() {
+		go scan()
+		for {
+			select {
+			case <-stop:
+				close(texts)
+				return
+			case <-time.After(timeout):
+				continue
+			case <-c:
+				go scan()
```

Comment:
If `c <- s.Scan()` happens before `texts <- s.Text()`, you will be advancing the scanner and skipping the current token, I think.

Reply:
Nice catch! You're right, I should swap `go scan()` and `texts <- s.Text()`. I also realized that I should return if `<- c` is `false`.

Reply:
Fixed [here](https://github.com/src-d/engine/pull/216/commits/af225a0d5b6dd8f84d9c6b97d78d93e54377c87b).